config/log_config.go

```diff
@@ -0,0 +1,77 @@
+package config
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"go.viam.com/rdk/logging"
+)
+
+// LogPatternConfig is an instance of a level specification for a given logger.
+type LogPatternConfig struct {
+	Pattern string `json:"pattern"`
+	Level   string `json:"level"`
+}
+
+func validatePattern(pattern string) bool {
+	size := len(pattern)
+	if size == 0 || pattern[0] == '.' || pattern[size-1] == '.' {
+		return false
+	}
+	for i := 1; i < size; i++ {
+		if pattern[i] == '.' && pattern[i-1] == '.' {
+			return false
+		}
+	}
+	r := regexp.MustCompile(`^[a-z_-]+$|^\*$`)
+	for _, sep := range strings.Split(pattern, ".") {
+		if !r.MatchString(sep) {
+			return false
+		}
+	}
+	return true
+}
+
+// UpdateLoggerRegistry updates the logger registry if necessary  with the specified logConfig.
+func UpdateLoggerRegistry(logConfig []LogPatternConfig, loggerRegistry map[string]logging.Logger) (map[string]logging.Logger, error) {
+	newLogRegistry := make(map[string]logging.Logger)
+
+	for _, lpc := range logConfig {
+		if !validatePattern(lpc.Pattern) {
+			return nil, errors.New("Failed to Validate a Pattern")
+		}
+
+		matcher := "^"
+		for _, ch := range lpc.Pattern {
+			if ch == '*' {
+				matcher += `(\w|\.)+`
+			} else {
+				matcher += string(ch)
+			}
+		}
+		matcher += "$"
+		r := regexp.MustCompile(matcher)
```

Comment:
RE: https://github.com/viamrobotics/rdk/pull/4173/files#r1670733048

The reason I asked why adding `\.` on the `'*'` case is necessary is because we add already `'.'` to the pattern on the else case. So if I'm reading the current code correctly, a pattern like `a.*.c` will be transformed into regexp `a.(\w|\.)+.c`. In this case, I _think_ we want to get regexp `a\.\w+\.c`.

```suggestion
		var matcher strings.Builder
		matcher.WriteString("^")
		for _, ch := range lpc.Pattern {
			switch ch {
			case '*':
				matcher.WriteString(`\w+`)
			case  '.':
				matcher.WriteString(`\.`)
			default:
				matcher.WriteRune(ch)
			}
		}
		matcher.WriteString("$")
		r := regexp.MustCompile(matcher.String())
```

Reply:
you are right that the `\.` isnt necessary for some of the `*` within the pattern string, but I believe it is necessary for when the last character of the pattern string is `*`. For instance, we want the pattern  `a.*` to match logger names such as `a.b.c`. If we converted the matcher to `^a.\w+$`, we would only match logger names of the form `a.x` because `\w` does not include the period character. Instead I think we want `^a.\w+(\.\w+)*` to match the case described.

Reply:
> I believe it is necessary for when the last character of the pattern string is *

Ah okay, I missed this requirement, so `a.*` should match `a.b` as well as `a.b.c`, etc